tools/stats/monitor.py

```diff
@@ -0,0 +1,104 @@
+    # when running this script in the background, pip installs the packages to
+    # a location that is not in a path python uses to look for modules, so we
+    # manually add the location to the path
+    location = subprocess.run(
+        [sys.executable, "-m", "pip", "show", package_name],
+        stderr=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+    )
+    for s in location.stdout.decode("utf-8").splitlines():
+        if "Location:" in s:
+            sys.path.append(s.split(" ")[1])
+
+
+def main() -> None:
+    pip_install("psutil")
+    pip_install("pynvml")
+    import psutil  # type: ignore[import]
+    import pynvml  # type: ignore[import]
+
+    def get_processes_running_python_tests() -> List[Any]:
+        python_processes = []
+        for process in psutil.process_iter():
+            try:
+                if "python" in process.name() and process.cmdline():
+                    python_processes.append(process)
+            except (psutil.NoSuchProcess, psutil.AccessDenied):
+                # access denied or the process died
+                pass
+        return python_processes
+
+    def get_per_process_cpu_info() -> List[Dict[str, Any]]:
+        processes = get_processes_running_python_tests()
+        per_process_info = []
+        for p in processes:
+            info = {
+                "pid": p.pid,
+                "cmd": " ".join(p.cmdline()),
+                "cpu_percent": p.cpu_percent(),
+                "rss_memory": p.memory_info().rss,
+                "uss_memory": p.memory_full_info().uss,
+            }
+            if "pss" in p.memory_full_info():
+                # only availiable in linux
+                info["pss_memory"] = p.memory_full_info().pss
+            per_process_info.append(info)
+        return per_process_info
+
+    def get_per_process_gpu_info(handle: Any) -> List[Dict[str, Any]]:
+        processes = pynvml.nvmlDeviceGetComputeRunningProcesses(handle)
+        per_process_info = []
+        for p in processes:
+            info = {"pid": p.pid, "gpu_memory": p.usedGpuMemory}
+            per_process_info.append(info)
+        return per_process_info
+
+    handle = None
+    try:
+        pynvml.nvmlInit()
+        handle = pynvml.nvmlDeviceGetHandleByIndex(0)
+    except pynvml.NVMLError:
+        # no pynvml avaliable, probably because not cuda
+        pass
+
+    while True:
+        try:
+            stats = {
+                "time": datetime.datetime.utcnow().isoformat("T") + "Z",
+                "total_cpu_percent": psutil.cpu_percent(),
+                "per_process_cpu_info": get_per_process_cpu_info(),
+            }
+            if handle is not None:
+                stats["per_process_gpu_info"] = get_per_process_gpu_info(handle)
+        except Exception as e:
+            stats = {
+                "time": datetime.datetime.utcnow().isoformat("T") + "Z",
+                "error": str(e),
+            }
+        finally:
+            print(json.dumps(stats))
```

Comment:
Minor: because [SIGTERM](https://stackoverflow.com/questions/9930576/python-what-is-the-default-handling-of-sigterm) is used to kill the process, I think there is a rare but potential issue with invalid JSON output (broken last line because the process is killed during the print statement). You could register a trivial SIGTERM handle and exit gracefully just to be sure

Reply:
done